### ISSUE TITLE
[BUG]: nan loss when using spectral ffn with v2

### ISSUE THREAD

**Issue by Rhys-McAlister** [chemprop/chemprop#544]
Hi, I've been attempting to use the spectralffn block and am getting nan for the training loss, my code for the model is setup as follows:

`
mp = message_passing.molecule.BondMessageBlock()
agg = nn.agg.MeanAggregation()
ffn = nn.readout.SpectralFFN(
    n_tasks=1801,
    spectral_activation='exp'
)
batch_norm = False
metric_list = [nn.metrics.SIDMetric(), nn.metrics.RMSEMetric(), nn.metrics.MAEMetric()] 
mpnn = model.MPNN(mp, agg, ffn, batch_norm, metric_list)
`

**Comment by Rhys-McAlister** [comment]
Any help with this would be greatly appreciated

**Comment by am2145** [comment]
Hi Rhys. 

I apologize for taking so long to look at this. In your spectral data, do you have any negative target values? I have had a look at  the SIDLoss function's definition and at the moment if it is used with no masking and negative values are present in the targets, the `.log()` operations would result in nans being present and might explain why you are encountering a nan loss during training. 

The original paper that implemented SIDLoss as a loss function used exponential activation and normalization of the spectra to constrain the predictions and targets to be positive. We may need to do something similar here and I'll look into this further.

**Comment by Rhys-McAlister** [comment]
Hello, thanks for the response.

I've tried the code again and made sure there was no negative values present in the target columns with the following setup for the model code:

mp = nn.BondMessagePassing(
    depth = 6,
    d_h = 2200,
    dropout=0.05
)
agg = nn.MeanAggregation()
ffn = nn.RegressionFFN(
    input_dim=2200,
    n_layers=6,
    hidden_dim=2200,
    dropout=0.05,
    # loc=scaler.mean_, # pass in the mean of the training targets
    # scale=scaler.scale_,
    n_tasks=1801 # pass in the scale of the training targets
)



```
batch_norm=False
metric_list = [nn.metrics.SIDLoss(), nn.metrics.RMSEMetric()] # Only the first metric is used for training and early stopping
mpnn = models.MPNN(mp, agg, ffn, batch_norm, metric_list)
```
When I attempt to use this code I get this error now:

`File [~\OneDrive\Documents\Github\chemprop\chemprop\nn\loss.py:70](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:70), in LossFunction.__call__(self, preds, targets, mask, w_s, w_t, lt_mask, gt_mask)
     [43](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:43) """Calculate the mean loss function value given predicted and target values
     [44](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:44) 
     [45](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:45) Parameters
   (...)
     [67](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:67)     a scalar containing the fully reduced loss
     [68](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:68) """
     [69](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:69) L = self.forward(preds, targets, mask, w_s, w_t, lt_mask, gt_mask)
---> [70](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:70) L = L * w_s.view(-1, 1) * w_t.view(1, -1) * mask
     [72](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/rhys-/OneDrive/Documents/Github/chemprop/notebooks/big_data/~/OneDrive/Documents/Github/chemprop/chemprop/nn/loss.py:72) return L.sum() / mask.sum()

AttributeError: 'NoneType' object has no attribute 'view'`

Additionally, when I go back to using RMSE as the loss function and SIDMetric I still receive nan values when calling `.test()` on the test set.

**Comment by Rhys-McAlister** [comment]
Thank you that is quite helpful, in the interim is there any modifications I can make to RegressionFFN to approximate the behaviour of the SpectralFFN?

**Comment by Rhys-McAlister** [comment]
> The error you are seeing is due to calling `SIDLoss()` in the metrics list instead of `SIDMetric()` in the metrics list. The metrics list in your original post should work, but do let me know if you see an issue there.
> 
> As for the nan values, this can happen if the ratio of preds/targets becomes negative as `SIDLoss()` uses log() functions. I can see this occuring for very small absorbances as `RegressionFFN()` doesn't enforce positivity of predictions. To address this, you can make use of the threshold parameter when declaring the metric. For example: `metrics.SIDMetric(threshold = 0.0001)` will clamp predicted values below the threshold to be 0.0001.
> 
> Another thing to note is that the original spectral prediction implementation enforces positivity for and normalizes both the target spectra and the predictions, so the losses and performance will likely be different when using `RegressionFFN()` as it does not do this. `SpectralFFN` is still a work in progress, but we are working to have an implementation soon that will handle these aspects and be applicable for your use case.
> 
> I hope this helps.

With the changes you mentioned implemented, the training loss returns an actual value now but the validation loss still returns nan, this issue persists across different datasets as well.

**Comment by JacksonBurns** [comment]
> With the changes you mentioned implemented, the training loss returns an actual value now but the validation loss still returns nan, this issue persists across different datasets as well.

@am2145 @kevingreenman @KnathanM if the loss function works for training but not validation, could this be a case of #380 (i.e. just a problem with nan-averaging)?

**Comment by kevingreenman** [comment]
I think #380 was a separate issue related to division by 0. The nan-averaging issues were in #453 and #458. I think this issue is different but might be somewhat related.

[`SIDMetric`](https://github.com/chemprop/chemprop/blob/327c31e932db8cea92e9d9a221c288eaeee911cb/chemprop/nn/metrics.py#L176)  inherits from [`SIDLoss`](https://github.com/chemprop/chemprop/blob/327c31e932db8cea92e9d9a221c288eaeee911cb/chemprop/nn/loss.py#L302) and [`Metric`](https://github.com/chemprop/chemprop/blob/327c31e932db8cea92e9d9a221c288eaeee911cb/chemprop/nn/metrics.py#L36). So I think the `__call__` method of `SIDMetric` is inherited from `Metric`, where it returns `self.forward(preds, targets, mask, lt_mask, gt_mask)[mask].mean()`, and the `forward` method is inherited from `SIDLoss`. Perhaps we need to change something about how `SIDLoss` is defined or override the `__call__` method?

**Comment by am2145** [comment]
An initial implementation of SpectralFFN() is now merged in v2 from #624. To summarize the discussion from there as well as v1's approach, both the targets and predictions need to be positive and sum-normalized for the SpectralFFN() to work as intended. Then, in the `SIDloss` function we will only normalize if a Boolean `self.check_inputs` is passed as true. I'll open separate issues and work on PRs for both of these, but this should hopefully help resolve the NaN issues.

**Comment by am2145** [comment]
Long overdue, but I am closing this issue and have opened new issues in #1008 and #1009 to implement what was discussed here.
